File: javascript/src/preset-tools.ts
import type { ListPresetsResponse, Preset } from "./types/catalog.js";
import type { ListToolsResponse, PublicTool, Tool } from "./types/tools.js";
import { validateUniqueToolNames } from "./tool-validation.js";

export interface PresetToolCatalogClient {
  presets: {
    list(): Promise<ListPresetsResponse>;
  };
  tools: {
    list(): Promise<ListToolsResponse>;
  };
}

export type UnknownPresetToolBehavior = "stub" | "omit" | "error";

export interface ResolvePresetToolsOptions {
  /**
   * Preset id whose policy.allowed_tools should be resolved.
   */
  preset: string;
  /**
   * Additional caller/client tools to append after the preset tools.
   */
  tools?: readonly Tool[];
  /**
   * Optional pre-fetched preset catalog. Use this when your app caches discovery
   * responses and wants deterministic request construction.
   */
  presets?: readonly Preset[];
  /**
   * Optional pre-fetched tool catalog.
   */
  toolCatalog?: readonly PublicTool[];
  /**
   * Behavior when a preset allows a tool that is not present in the supplied
   * or fetched tool catalog. "stub" keeps a name-only tool definition so the
   * backend can still enrich it by name.
   */
  unknownPresetTool?: UnknownPresetToolBehavior;
}

export interface ResolvePresetToolsResult {
  preset: Preset;
  tools: Tool[];
  missingToolNames: string[];
}

/**
 * Resolve a preset's allowed tool names into concrete request tools and append
 * caller-provided tools. This is a convenience for hybrid apps that need to add
 * local/client tools while preserving the preset's default server tools.
 *
 * The Agent API request surface remains OpenAI-compatible: the returned array
 * is intended for the normal CreateResponseRequest.tools field.
 */
export async function resolvePresetTools(
  client: PresetToolCatalogClient,
  options: ResolvePresetToolsOptions,
): Promise<ResolvePresetToolsResult> {
  const presets = options.presets ?? (await client.presets.list()).data;
  const toolCatalog = options.toolCatalog ?? (await client.tools.list()).data;
  return resolvePresetToolsFromCatalog({ ...options, presets, toolCatalog });
}

export function resolvePresetToolsFromCatalog(options: ResolvePresetToolsOptions): ResolvePresetToolsResult {
  const presetId = options.preset.trim();
  if (!presetId) {
    throw new Error("preset is required");
  }
  const preset = options.presets?.find((row) => row.preset === presetId);
  if (!preset) {
    throw new Error(`preset not found: ${presetId}`);
  }

  const catalogByName = new Map<string, PublicTool>();
  for (const tool of options.toolCatalog ?? []) {
    const name = tool.name?.trim();
    if (name) catalogByName.set(name, tool);
  }

  const missingToolNames: string[] = [];
  const presetTools: Tool[] = [];
  const unknownPresetTool = options.unknownPresetTool ?? "stub";

  for (const name of preset.policy?.allowed_tools ?? []) {
    const trimmed = name.trim();
    if (!trimmed) continue;
    const catalogTool = catalogByName.get(trimmed);
    if (catalogTool) {
      presetTools.push(publicToolToRequestTool(catalogTool));
      continue;
    }
    missingToolNames.push(trimmed);
    if (unknownPresetTool === "error") {
      throw new Error(`preset tool not found in catalog: ${trimmed}`);
    }
    if (unknownPresetTool === "stub") {
      presetTools.push({ name: trimmed } as Tool);
    }
  }

  return {
    preset,
    tools: mergeTools(presetTools, options.tools ?? []),
    missingToolNames,
  };
}

export function mergeTools(...groups: Array<readonly Tool[]>): Tool[] {
  const out: Tool[] = [];
  for (const group of groups) {
    for (const tool of group) {
      const name = tool.name?.trim();
      if (!name) {
        throw new Error("tools[].name is required");
      }
      out.push({ ...tool, name } as Tool);
    }
  }
  validateUniqueToolNames(out);
  return out;
}

export function publicToolToRequestTool(tool: PublicTool): Tool {
  const out: Record<string, unknown> = { name: tool.name };
  if (tool.type) out.type = tool.type;
  if (tool.description) out.description = tool.description;
  if (tool.parameters) out.parameters = tool.parameters;
  if (tool.max_tokens != null) out.max_tokens = tool.max_tokens;
  if (tool.max_tokens_per_page != null) out.max_tokens_per_page = tool.max_tokens_per_page;
  if (tool.version) out.version = tool.version;
  return out as unknown as Tool;
}

File: python/src/agent_api/preset_tools.py
from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Any, Literal, Protocol, cast

from agent_api.tool_validation import validate_unique_tool_names
from agent_api.types.responses import Preset, PublicTool
from agent_api.types.tools import Tool

UnknownPresetToolBehavior = Literal["stub", "omit", "error"]


class _ListResource(Protocol):
    def list(self) -> dict[str, Any]: ...


class _AsyncListResource(Protocol):
    async def list(self) -> dict[str, Any]: ...


class PresetToolCatalogClient(Protocol):
    presets: _ListResource
    tools: _ListResource


class AsyncPresetToolCatalogClient(Protocol):
    presets: _AsyncListResource
    tools: _AsyncListResource


@dataclass(frozen=True)
class ResolvePresetToolsResult:
    preset: Preset
    tools: list[Tool]
    missing_tool_names: list[str]


def resolve_preset_tools(
    client: PresetToolCatalogClient,
    *,
    preset: str,
    tools: Sequence[Tool] | None = None,
    presets: Sequence[Preset] | None = None,
    tool_catalog: Sequence[PublicTool] | None = None,
    unknown_preset_tool: UnknownPresetToolBehavior = "stub",
) -> ResolvePresetToolsResult:
    """Resolve preset allowed tools and append caller/client tools.

    Use this for hybrid apps that need to add local function tools while preserving
    the server-side tools provided by a preset. The returned list is intended for
    the normal CreateResponseRequest.tools field.
    """

    resolved_presets = presets if presets is not None else cast(list[Preset], client.presets.list().get("data", []))
    resolved_tools = tool_catalog if tool_catalog is not None else cast(list[PublicTool], client.tools.list().get("data", []))
    return resolve_preset_tools_from_catalog(
        preset=preset,
        tools=tools,
        presets=resolved_presets,
        tool_catalog=resolved_tools,
        unknown_preset_tool=unknown_preset_tool,
    )


async def async_resolve_preset_tools(
    client: AsyncPresetToolCatalogClient,
    *,
    preset: str,
    tools: Sequence[Tool] | None = None,
    presets: Sequence[Preset] | None = None,
    tool_catalog: Sequence[PublicTool] | None = None,
    unknown_preset_tool: UnknownPresetToolBehavior = "stub",
) -> ResolvePresetToolsResult:
    resolved_presets = presets if presets is not None else cast(list[Preset], (await client.presets.list()).get("data", []))
    resolved_tools = tool_catalog if tool_catalog is not None else cast(list[PublicTool], (await client.tools.list()).get("data", []))
    return resolve_preset_tools_from_catalog(
        preset=preset,
        tools=tools,
        presets=resolved_presets,
        tool_catalog=resolved_tools,
        unknown_preset_tool=unknown_preset_tool,
    )


def resolve_preset_tools_from_catalog(
    *,
    preset: str,
    tools: Sequence[Tool] | None = None,
    presets: Sequence[Preset] | None = None,
    tool_catalog: Sequence[PublicTool] | None = None,
    unknown_preset_tool: UnknownPresetToolBehavior = "stub",
) -> ResolvePresetToolsResult:
    preset_id = preset.strip()
    if not preset_id:
        raise ValueError("preset is required")

    matched = next((row for row in presets or [] if row.get("preset") == preset_id), None)
    if matched is None:
        raise ValueError(f"preset not found: {preset_id}")

    catalog_by_name = {name: tool for tool in tool_catalog or [] if (name := str(tool.get("name", "")).strip())}
    missing_tool_names: list[str] = []
    preset_tools: list[Tool] = []

    for name in matched.get("policy", {}).get("allowed_tools", []):
        trimmed = str(name).strip()
        if not trimmed:
            continue
        catalog_tool = catalog_by_name.get(trimmed)
        if catalog_tool is not None:
            preset_tools.append(public_tool_to_request_tool(catalog_tool))
            continue
        missing_tool_names.append(trimmed)
        if unknown_preset_tool == "error":
            raise ValueError(f"preset tool not found in catalog: {trimmed}")
        if unknown_preset_tool == "stub":
            preset_tools.append(cast(Tool, {"name": trimmed}))

    return ResolvePresetToolsResult(
        preset=matched,
        tools=merge_tools(preset_tools, tools or []),
        missing_tool_names=missing_tool_names,
    )


def merge_tools(*groups: Sequence[Tool]) -> list[Tool]:
    out: list[Tool] = []
    for group in groups:
        for tool in group:
            name = str(tool.get("name", "")).strip()
            if not name:
                raise ValueError("tools[].name is required")
            normalized = dict(tool)
            normalized["name"] = name
            out.append(cast(Tool, normalized))
    validate_unique_tool_names(out)
    return out


def public_tool_to_request_tool(tool: PublicTool) -> Tool:
    out: dict[str, Any] = {"name": tool["name"]}
    for key in ("type", "description", "parameters", "max_tokens", "max_tokens_per_page", "version"):
        if key in tool:
            out[key] = tool[key]  # type: ignore[literal-required]
    return cast(Tool, out)
